File: src/shesasyncer/consensus/confidence.py
from statistics import mean
from ..core.models import AlignmentEvidence


def confidence(evidence: list[AlignmentEvidence]) -> float:
    if not evidence:
        return 0.0
    return round(mean(max(0.0, min(1.0, e.timing.confidence)) for e in evidence), 4)


def conflicts(evidence: list[AlignmentEvidence], tolerance: float = 0.75) -> list[int]:
    """Return line indexes where independent evidence sources disagree."""
    by_line: dict[int, list[AlignmentEvidence]] = {}
    for item in evidence:
        by_line.setdefault(item.line_index, []).append(item)
    result = []
    for index, items in by_line.items():
        starts = [x.timing.start for x in items]
        ends = [x.timing.end for x in items]
        if len(items) > 1 and (max(starts) - min(starts) > tolerance or max(ends) - min(ends) > tolerance):
            result.append(index)
    return result

File: src/shesasyncer/consensus/confidence.js
export function confidence(evidence) {
  if (!evidence.length) return 0;
  return Number((evidence.reduce((s, e) => s + Math.max(0, Math.min(1, e.timing.confidence)), 0) / evidence.length).toFixed(4));
}

export function conflicts(evidence, tolerance = 0.75) {
  const byLine = new Map();
  for (const item of evidence) { if (!byLine.has(item.lineIndex)) byLine.set(item.lineIndex, []); byLine.get(item.lineIndex).push(item); }
  const result = [];
  for (const [index, items] of byLine) {
    const starts = items.map(x => x.timing.start), ends = items.map(x => x.timing.end);
    if (items.length > 1 && (Math.max(...starts) - Math.min(...starts) > tolerance || Math.max(...ends) - Math.min(...ends) > tolerance)) result.push(index);
  }
  return result;
}
